File: app/src/test/java/org/alkaline/taskbrain/ui/currentnote/undo/UnifiedUndoManagerTest.kt
package org.alkaline.taskbrain.ui.currentnote.undo

import org.alkaline.taskbrain.ui.currentnote.EditorController
import org.alkaline.taskbrain.ui.currentnote.EditorState
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test

class UnifiedUndoManagerTest {

    private lateinit var unified: UnifiedUndoManager

    @Before
    fun setUp() {
        unified = UnifiedUndoManager()
    }

    /** Creates an editor pair, registers it, sets baseline with initial text. */
    private fun createEditor(
        contextId: String,
        initialText: String = "initial"
    ): Pair<EditorState, EditorController> {
        val state = EditorState()
        val controller = EditorController(state)
        state.updateFromText(initialText)
        controller.undoManager.setBaseline(state)
        unified.registerEditor(contextId, controller)
        return state to controller
    }

    /** Simulates an edit: update text, mark changed, commit. */
    private fun edit(state: EditorState, controller: EditorController, newText: String) {
        controller.undoManager.beginEditingLine(state, 0)
        state.updateFromText(newText)
        controller.undoManager.markContentChanged()
        controller.commitUndoState()
    }

    // ==================== Register / Unregister ====================

    @Test
    fun `registerEditor hooks callbacks`() {
        val (state, controller) = createEditor("main")

        edit(state, controller, "edited")

        assertTrue(unified.canUndo)
    }

    @Test
    fun `unregisterEditor removes callbacks`() {
        val (state, controller) = createEditor("main")

        unified.unregisterEditor("main")

        // Edit after unregister should not add to unified stack
        edit(state, controller, "edited")

        assertFalse(unified.canUndo)
    }

    // ==================== Chronological Undo Order ====================

    @Test
    fun `undo follows chronological order across editors`() {
        val (stateMain, ctrlMain) = createEditor("main", "main-v0")
        val (stateA, ctrlA) = createEditor("inlineA", "a-v0")

        // Edit main first, then inline A
        edit(stateMain, ctrlMain, "main-v1")
        edit(stateA, ctrlA, "a-v1")

        val activations = mutableListOf<String>()

        // First undo should target inline A (most recent)
        val result1 = unified.undo("main") { activations.add(it) }
        assertNotNull(result1)
        assertEquals("inlineA", result1!!.contextId)
        assertEquals(listOf("a-v0"), result1.snapshot.lineContents)

        // Second undo should target main
        val result2 = unified.undo("inlineA") { activations.add(it) }
        assertNotNull(result2)
        assertEquals("main", result2!!.contextId)
        assertEquals(listOf("main-v0"), result2.snapshot.lineContents)
    }

    @Test
    fun `undo returns null when nothing to undo`() {
        createEditor("main", "text")

        val result = unified.undo("main") { }
        assertNull(result)
    }

    // ==================== Redo Order ====================

    @Test
    fun `redo restores in correct order after undo`() {
        val (stateMain, ctrlMain) = createEditor("main", "main-v0")
        val (stateA, ctrlA) = createEditor("inlineA", "a-v0")

        edit(stateMain, ctrlMain, "main-v1")
        edit(stateA, ctrlA, "a-v1")

        // Undo both
        val undo1 = unified.undo("inlineA") { }
        assertNotNull(undo1)
        // After undo, stateA still has "a-v1" text (undo returns snapshot but
        // UnifiedUndoManager delegates restore to caller). For redo to work,
        // simulate that the caller restored the state.
        stateA.updateFromText("a-v0")

        val undo2 = unified.undo("main") { }
        assertNotNull(undo2)
        stateMain.updateFromText("main-v0")

        // Redo should restore main first (it was undone last)
        val redo1 = unified.redo("main") { }
        assertNotNull(redo1)
        assertEquals("main", redo1!!.contextId)

        // Redo should restore inline A next
        stateMain.updateFromText("main-v1")
        val redo2 = unified.redo("main") { }
        assertNotNull(redo2)
        assertEquals("inlineA", redo2!!.contextId)
    }

    @Test
    fun `canRedo is false initially`() {
        createEditor("main")
        assertFalse(unified.canRedo)
    }

    @Test
    fun `canRedo is true after undo`() {
        val (state, controller) = createEditor("main", "v0")
        edit(state, controller, "v1")

        unified.undo("main") { }

        assertTrue(unified.canRedo)
    }

    @Test
    fun `canRedo is false without undo`() {
        val (state, controller) = createEditor("main", "v0")
        edit(state, controller, "v1")

        assertFalse(unified.canRedo)
    }

    // ==================== activateEditor Callback ====================

    @Test
    fun `activateEditor is called when undo targets different editor`() {
        val (stateMain, ctrlMain) = createEditor("main", "main-v0")
        val (stateA, ctrlA) = createEditor("inlineA", "a-v0")

        edit(stateA, ctrlA, "a-v1")

        val activations = mutableListOf<String>()

        // Active is "main" but undo target is "inlineA"
        unified.undo("main") { activations.add(it) }

        assertEquals(listOf("inlineA"), activations)
    }

    @Test
    fun `activateEditor is not called when undo targets same editor`() {
        val (state, controller) = createEditor("main", "v0")
        edit(state, controller, "v1")

        val activations = mutableListOf<String>()

        unified.undo("main") { activations.add(it) }

        assertTrue(activations.isEmpty())
    }

    @Test
    fun `activateEditor is called on redo targeting different editor`() {
        val (stateMain, ctrlMain) = createEditor("main", "main-v0")
        val (stateA, ctrlA) = createEditor("inlineA", "a-v0")

        edit(stateA, ctrlA, "a-v1")

        // Undo from inlineA context (no activation needed)
        unified.undo("inlineA") { }
        stateA.updateFromText("a-v0")

        val activations = mutableListOf<String>()

        // Redo from main context - should activate inlineA
        unified.redo("main") { activations.add(it) }

        assertEquals(listOf("inlineA"), activations)
    }

    // ==================== abandonHistory ====================

    @Test
    fun `abandonHistory clears undo stack`() {
        val (state, controller) = createEditor("main", "v0")
        edit(state, controller, "v1")

        assertTrue(unified.canUndo)

        unified.abandonHistory()

        assertFalse(unified.canUndo)
    }

    @Test
    fun `abandonHistory clears redo stack`() {
        val (state, controller) = createEditor("main", "v0")
        edit(state, controller, "v1")

        unified.undo("main") { }
        assertTrue(unified.canRedo)

        unified.abandonHistory()

        assertFalse(unified.canRedo)
    }

    // ==================== canUndo with uncommitted changes ====================

    @Test
    fun `canUndo is true when editor has uncommitted changes`() {
        val (state, controller) = createEditor("main", "v0")

        // Start editing but don't commit
        controller.undoManager.beginEditingLine(state, 0)
        state.updateFromText("v1")
        controller.undoManager.markContentChanged()

        assertTrue(unified.canUndo)
    }

    @Test
    fun `canUndo is false when no editors have changes`() {
        createEditor("main", "text")
        createEditor("inlineA", "text")

        assertFalse(unified.canUndo)
    }

    // ==================== New edits clear redo ====================

    @Test
    fun `new edit clears unified redo stack`() {
        val (state, controller) = createEditor("main", "v0")
        edit(state, controller, "v1")

        unified.undo("main") { }
        assertTrue(unified.canRedo)

        // Restore state after undo, then make a new edit
        state.updateFromText("v0")
        edit(state, controller, "v2")

        assertFalse(unified.canRedo)
    }

    // ==================== reset ====================

    @Test
    fun `reset clears everything and unregisters editors`() {
        val (state, controller) = createEditor("main", "v0")
        edit(state, controller, "v1")

        unified.reset()

        assertFalse(unified.canUndo)
        assertFalse(unified.canRedo)

        // Callbacks should be disconnected - new edits should not affect unified manager
        edit(state, controller, "v2")
        assertFalse(unified.canUndo)
    }

    // ==================== stateVersion ====================

    @Test
    fun `stateVersion increments on undo`() {
        val (state, controller) = createEditor("main", "v0")
        edit(state, controller, "v1")

        val versionBefore = unified.stateVersion
        unified.undo("main") { }

        assertTrue(unified.stateVersion > versionBefore)
    }

    @Test
    fun `stateVersion increments on abandonHistory`() {
        val versionBefore = unified.stateVersion
        unified.abandonHistory()
        assertTrue(unified.stateVersion > versionBefore)
    }
}

File: web/src/__tests__/editor/UnifiedUndoManager.test.ts
import { describe, it, expect } from 'vitest'
import { EditorState } from '../../editor/EditorState'
import { EditorController } from '../../editor/EditorController'
import { UndoManager } from '../../editor/UndoManager'
import { UnifiedUndoManager } from '../../editor/UnifiedUndoManager'

function createEditor(): EditorController {
  const state = new EditorState()
  const undoManager = new UndoManager()
  return new EditorController(state, undoManager)
}

/**
 * Simulate a user edit: update text, mark changed, then commit.
 * This creates a real undo entry on the editor's UndoManager.
 */
function simulateEdit(controller: EditorController, newText: string): void {
  // Begin editing so a pending snapshot is captured
  controller.undoManager.beginEditingLine(
    controller.state.lines,
    controller.state.focusedLineIndex,
    controller.state.focusedLineIndex,
  )
  // Apply the text change
  controller.state.updateFromText(newText)
  // Mark that content has changed (enables undo)
  controller.undoManager.markContentChanged()
  // Commit the pending state (pushes to undo stack)
  controller.commitUndoState()
}

describe('UnifiedUndoManager initial state', () => {
  it('canUndo is false with no editors', () => {
    const unified = new UnifiedUndoManager()
    expect(unified.canUndo).toBe(false)
  })

  it('canRedo is false with no editors', () => {
    const unified = new UnifiedUndoManager()
    expect(unified.canRedo).toBe(false)
  })

  it('stateVersion starts at 0', () => {
    const unified = new UnifiedUndoManager()
    expect(unified.stateVersion).toBe(0)
  })
})

describe('UnifiedUndoManager register/unregister', () => {
  it('registerEditor sets up callbacks on the UndoManager', () => {
    const unified = new UnifiedUndoManager()
    const controller = createEditor()

    unified.registerEditor('main', controller)
    expect(controller.undoManager.onEntryPushed).not.toBeNull()
    expect(controller.undoManager.onRedoCleared).not.toBeNull()
  })

  it('unregisterEditor clears callbacks', () => {
    const unified = new UnifiedUndoManager()
    const controller = createEditor()

    unified.registerEditor('main', controller)
    unified.unregisterEditor('main')
    expect(controller.undoManager.onEntryPushed).toBeNull()
    expect(controller.undoManager.onRedoCleared).toBeNull()
  })

  it('unregisterEditor is safe for non-existent ID', () => {
    const unified = new UnifiedUndoManager()
    expect(() => unified.unregisterEditor('nonexistent')).not.toThrow()
  })
})

describe('UnifiedUndoManager canUndo', () => {
  it('returns true when a registered editor has uncommitted changes', () => {
    const unified = new UnifiedUndoManager()
    const controller = createEditor()
    unified.registerEditor('main', controller)

    // Begin editing and mark changed but do NOT commit
    controller.undoManager.beginEditingLine(
      controller.state.lines,
      controller.state.focusedLineIndex,
      controller.state.focusedLineIndex,
    )
    controller.state.updateFromText('some text')
    controller.undoManager.markContentChanged()

    expect(unified.canUndo).toBe(true)
  })

  it('returns true when unified undo stack has entries', () => {
    const unified = new UnifiedUndoManager()
    const controller = createEditor()
    unified.registerEditor('main', controller)

    simulateEdit(controller, 'hello')
    expect(unified.canUndo).toBe(true)
  })

  it('returns false when no edits have been made', () => {
    const unified = new UnifiedUndoManager()
    const controller = createEditor()
    unified.registerEditor('main', controller)

    expect(unified.canUndo).toBe(false)
  })
})

describe('UnifiedUndoManager canRedo', () => {
  it('returns false initially', () => {
    const unified = new UnifiedUndoManager()
    const controller = createEditor()
    unified.registerEditor('main', controller)
    expect(unified.canRedo).toBe(false)
  })

  it('returns true after undo', () => {
    const unified = new UnifiedUndoManager()
    const controller = createEditor()
    unified.registerEditor('main', controller)

    simulateEdit(controller, 'hello')
    unified.undo('main', () => {})

    expect(unified.canRedo).toBe(true)
  })
})

describe('UnifiedUndoManager chronological undo across editors', () => {
  it('undo targets the most recently edited editor', () => {
    const unified = new UnifiedUndoManager()
    const mainCtrl = createEditor()
    const inlineCtrl = createEditor()

    unified.registerEditor('main', mainCtrl)
    unified.registerEditor('inline-A', inlineCtrl)

    // Edit main first, then inline-A
    simulateEdit(mainCtrl, 'main edit')
    simulateEdit(inlineCtrl, 'inline edit')

    // First undo should target inline-A (most recent)
    const result1 = unified.undo('inline-A', () => {})
    expect(result1).not.toBeNull()
    expect(result1!.contextId).toBe('inline-A')

    // Second undo should target main
    const activatedEditors: string[] = []
    const result2 = unified.undo('inline-A', (id) => activatedEditors.push(id))
    expect(result2).not.toBeNull()
    expect(result2!.contextId).toBe('main')
    expect(activatedEditors).toContain('main')
  })

  it('redo restores in correct order after undo', () => {
    const unified = new UnifiedUndoManager()
    const mainCtrl = createEditor()
    const inlineCtrl = createEditor()

    unified.registerEditor('main', mainCtrl)
    unified.registerEditor('inline-A', inlineCtrl)

    simulateEdit(mainCtrl, 'main edit')
    simulateEdit(inlineCtrl, 'inline edit')

    // Undo both
    unified.undo('inline-A', () => {})
    unified.undo('inline-A', () => {})

    // Redo should replay in chronological order: main first, then inline-A
    const redo1 = unified.redo('main', () => {})
    expect(redo1).not.toBeNull()
    expect(redo1!.contextId).toBe('main')

    const redo2 = unified.redo('main', () => {})
    expect(redo2).not.toBeNull()
    expect(redo2!.contextId).toBe('inline-A')
  })
})

describe('UnifiedUndoManager activateEditor callback', () => {
  it('is invoked when undo targets a different editor', () => {
    const unified = new UnifiedUndoManager()
    const mainCtrl = createEditor()
    const inlineCtrl = createEditor()

    unified.registerEditor('main', mainCtrl)
    unified.registerEditor('inline-A', inlineCtrl)

    simulateEdit(mainCtrl, 'main edit')

    const activated: string[] = []
    // Undo from inline-A context, should switch to main
    unified.undo('inline-A', (id) => activated.push(id))
    expect(activated).toEqual(['main'])
  })

  it('is NOT invoked when undo targets the same editor', () => {
    const unified = new UnifiedUndoManager()
    const mainCtrl = createEditor()
    unified.registerEditor('main', mainCtrl)

    simulateEdit(mainCtrl, 'main edit')

    const activated: string[] = []
    unified.undo('main', (id) => activated.push(id))
    expect(activated).toEqual([])
  })

  it('is invoked for redo targeting a different editor', () => {
    const unified = new UnifiedUndoManager()
    const mainCtrl = createEditor()
    const inlineCtrl = createEditor()

    unified.registerEditor('main', mainCtrl)
    unified.registerEditor('inline-A', inlineCtrl)

    simulateEdit(inlineCtrl, 'inline edit')
    unified.undo('inline-A', () => {})

    const activated: string[] = []
    // Redo from main context, should switch to inline-A
    unified.redo('main', (id) => activated.push(id))
    expect(activated).toEqual(['inline-A'])
  })
})

describe('UnifiedUndoManager abandonHistory', () => {
  it('clears both undo and redo stacks', () => {
    const unified = new UnifiedUndoManager()
    const controller = createEditor()
    unified.registerEditor('main', controller)

    simulateEdit(controller, 'edit 1')
    simulateEdit(controller, 'edit 2')
    unified.undo('main', () => {})

    expect(unified.canUndo).toBe(true)
    expect(unified.canRedo).toBe(true)

    unified.abandonHistory()

    expect(unified.canUndo).toBe(false)
    expect(unified.canRedo).toBe(false)
  })

  it('increments stateVersion', () => {
    const unified = new UnifiedUndoManager()
    const before = unified.stateVersion
    unified.abandonHistory()
    expect(unified.stateVersion).toBeGreaterThan(before)
  })
})

describe('UnifiedUndoManager reset', () => {
  it('clears stacks, editors, and callbacks', () => {
    const unified = new UnifiedUndoManager()
    const controller = createEditor()
    unified.registerEditor('main', controller)

    simulateEdit(controller, 'edit')

    unified.reset()

    expect(unified.canUndo).toBe(false)
    expect(unified.canRedo).toBe(false)
    expect(controller.undoManager.onEntryPushed).toBeNull()
    expect(controller.undoManager.onRedoCleared).toBeNull()
  })
})

describe('UnifiedUndoManager stateVersion tracking', () => {
  it('increments on edit commit (via onEntryPushed)', () => {
    const unified = new UnifiedUndoManager()
    const controller = createEditor()
    unified.registerEditor('main', controller)

    const before = unified.stateVersion
    simulateEdit(controller, 'hello')
    expect(unified.stateVersion).toBeGreaterThan(before)
  })

  it('increments on undo', () => {
    const unified = new UnifiedUndoManager()
    const controller = createEditor()
    unified.registerEditor('main', controller)

    simulateEdit(controller, 'hello')
    const before = unified.stateVersion
    unified.undo('main', () => {})
    expect(unified.stateVersion).toBeGreaterThan(before)
  })

  it('increments on redo', () => {
    const unified = new UnifiedUndoManager()
    const controller = createEditor()
    unified.registerEditor('main', controller)

    simulateEdit(controller, 'hello')
    unified.undo('main', () => {})
    const before = unified.stateVersion
    unified.redo('main', () => {})
    expect(unified.stateVersion).toBeGreaterThan(before)
  })
})

describe('UnifiedUndoManager edge cases', () => {
  it('undo returns null when stack is empty', () => {
    const unified = new UnifiedUndoManager()
    const controller = createEditor()
    unified.registerEditor('main', controller)

    const result = unified.undo('main', () => {})
    expect(result).toBeNull()
  })

  it('redo returns null when redo stack is empty', () => {
    const unified = new UnifiedUndoManager()
    const controller = createEditor()
    unified.registerEditor('main', controller)

    const result = unified.redo('main', () => {})
    expect(result).toBeNull()
  })

  it('undo returns null when target editor was unregistered', () => {
    const unified = new UnifiedUndoManager()
    const controller = createEditor()
    unified.registerEditor('main', controller)

    simulateEdit(controller, 'edit')
    unified.unregisterEditor('main')

    // The unified stack still has the entry, but the editor is gone
    const result = unified.undo('main', () => {})
    expect(result).toBeNull()
  })

  it('new edit after undo clears unified redo stack', () => {
    const unified = new UnifiedUndoManager()
    const controller = createEditor()
    unified.registerEditor('main', controller)

    simulateEdit(controller, 'edit 1')
    unified.undo('main', () => {})
    expect(unified.canRedo).toBe(true)

    // A new edit should clear redo on both per-editor and unified level
    simulateEdit(controller, 'edit 2')
    expect(unified.canRedo).toBe(false)
  })
})
